lib/Pipeline/Contract.cpp

```diff
@@ -34,6 +34,21 @@ void Contract::deduceResults(ContainerToTargetsMap &StepStatus,
   deduceResults(StepStatus, OutputContainerTarget, Names);
 }
 
+bool Contract::satisfiedDeeperGranularityRequest(const Target &In) const {
+  // If target kind has depth equal or less than the kind of the input this
+  // cycle is not executed, otherwise it ensures that all granularities between
+  // the kind of the target and the kind of the input are *
+  if (TargetKind) {
+    for (size_t I = TargetKind->depth(); I < In.getKind().depth(); I++) {
+      if (In.getPathComponents()[I].isSingle()) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
```

Comment:
This is a bit obscure. Will this disappear if we get rid of `*` in targets?

If not, can you describe from an high level the problem?

Reply:
unfortunately removing * will not help, if anything this will become even more obscure because it will have to check if all elements are present.

this is just checking that if we are going from function to root then all functions must be present.

Reply:
i opted for the following comment
```
/**
 * returns false iff the output kind has less depth than the input and the input
 * is not *
 */

```

Reply:
OK (without C-style comment `:)`).

Can we turn this into `bool Target::isCompatibleWith(Kind *Other)`? Maybe we can have a different name, but this looks like a `Targets`' method.

Reply:
what do you mean c style comment? it is intended to be documentation.

it is a private method not exposed.

we can't have a compatibility thing, because it depends if it forward or backward and a couple of other things too

Reply:
Doxygen documentation in C++ uses `///`.

> we can't have a compatibility thing, because it depends if it forward or backward and a couple of other things too

Mmmh, you sure you can't make it work with `isCompatibleWith`?
`satisfiedDeeperGranularityRequest` is a very obscure name, we dropped the *granularity* term. Maybe `Contract::targetCanBePromotedTo`/`Contract::targetCanBeDemotedTo`?

Reply:
targetCanBePromotedToShallowerRank?

Reply:
`Shallower` is a big judgemental but works.

Reply:
yeah but promote demote is unclear it it means to lower rank or higher, this is clear that means lower